From e2f0452fe7019439dff9aaa4457c1f35d72179af Mon Sep 17 00:00:00 2001
From: Julia Reynolds <juliacr@google.com>
Date: Wed, 7 May 2025 12:47:47 -0400
Subject: [PATCH 2/2] Update isSameApp check

Only check uid, without a sandbox sdk exception

This small change involves a major test change since we previously
too permissive in our test setup

Test: NotificationManagerServiceTest
Test: HSUM NotificationManagerServiceTest
Test: SdkSandboxRestrictionsTest
Bug: 396667508
Fixes: 317957802
Flag: EXEMPT bug fix

Change-Id: I2be1537803a83cf44cc605f2850a3bb705b9bf32
(cherry picked from commit f5605ae25b15e1f8295e86c431d304e333ce56a4)
---
 .../NotificationManagerService.java           |  15 +-
 .../com/android/server/UiServiceTestCase.java |  18 +-
 .../NotificationManagerServiceTest.java       | 337 ++++++++++--------
 .../notification/PreferencesHelperTest.java   |  79 ++--
 4 files changed, 265 insertions(+), 184 deletions(-)

diff --git a/services/core/java/com/android/server/notification/NotificationManagerService.java b/services/core/java/com/android/server/notification/NotificationManagerService.java
index 3f2af4e24d6f..da76c82e76fc 100644
--- a/services/core/java/com/android/server/notification/NotificationManagerService.java
+++ b/services/core/java/com/android/server/notification/NotificationManagerService.java
@@ -4329,6 +4329,9 @@ public class NotificationManagerService extends SystemService {
          */
         @Override
         public boolean areNotificationsEnabledForPackage(String pkg, int uid) {
+            if (Process.isSdkSandboxUid(uid)) {
+                return false;
+            }
             enforceSystemOrSystemUIOrSamePackage(pkg,
                     "Caller not system or systemui or same package");
             if (UserHandle.getCallingUserId() != UserHandle.getUserId(uid)) {
@@ -4905,8 +4908,8 @@ public class NotificationManagerService extends SystemService {
         public NotificationChannel getConversationNotificationChannel(String callingPkg, int userId,
                 String targetPkg, String channelId, boolean returnParentIfNoConversationChannel,
                 String conversationId) {
-            if (canNotifyAsPackage(callingPkg, targetPkg, userId)
-                    || isCallerSystemOrSystemUiOrShell()) {
+            if (isCallerSystemOrSystemUiOrShell()
+                    || canNotifyAsPackage(callingPkg, targetPkg, userId)) {
                 int targetUid = INVALID_UID;
                 try {
                     targetUid = mPackageManagerClient.getPackageUidAsUser(targetPkg, userId);
@@ -6281,7 +6284,7 @@ public class NotificationManagerService extends SystemService {
                 String wellbeingPackage = getContext().getResources().getString(
                         com.android.internal.R.string.config_systemWellbeing);
                 boolean isCallerWellbeing = !TextUtils.isEmpty(wellbeingPackage)
-                        && mPackageManagerInternal.isSameApp(wellbeingPackage, uid, userId);
+                        && isCallerSameApp(wellbeingPackage, uid, userId);
                 if (!isCallerWellbeing) {
                     throw new IllegalArgumentException(
                             "Only the 'Wellbeing' package can use AutomaticZenRules with "
@@ -6571,7 +6574,7 @@ public class NotificationManagerService extends SystemService {
                     ? mEffectsSuppressors.get(0)
                     : null;
             if (isCallerSystemOrSystemUiOrShell() || suppressor == null
-                    || mPackageManagerInternal.isSameApp(suppressor.getPackageName(),
+                    || isCallerSameApp(suppressor.getPackageName(),
                     Binder.getCallingUid(), UserHandle.getUserId(Binder.getCallingUid()))) {
                 return suppressor;
             }
@@ -9025,7 +9028,7 @@ public class NotificationManagerService extends SystemService {
     int resolveNotificationUid(String callingPkg, String targetPkg, int callingUid, int userId)
             throws NameNotFoundException {
         if (userId == USER_ALL) {
-            userId = USER_SYSTEM;
+            userId = UserHandle.getUserId(callingUid);
         }
         // posted from app A on behalf of app A
         if (isCallerSameApp(targetPkg, callingUid, userId)
@@ -11864,7 +11867,7 @@ public class NotificationManagerService extends SystemService {
         if (uid == Process.ROOT_UID && ROOT_PKG.equals(pkg)) {
             return;
         }
-        if (!mPackageManagerInternal.isSameApp(pkg, uid, userId)) {
+        if (!UserHandle.isSameApp(uid, mPackageManagerInternal.getPackageUid(pkg, 0L, userId))) {
             throw new SecurityException("Package " + pkg + " is not owned by uid " + uid);
         }
     }
diff --git a/services/tests/uiservicestests/src/com/android/server/UiServiceTestCase.java b/services/tests/uiservicestests/src/com/android/server/UiServiceTestCase.java
index c9d5241c57b7..53d25ab6d7e9 100644
--- a/services/tests/uiservicestests/src/com/android/server/UiServiceTestCase.java
+++ b/services/tests/uiservicestests/src/com/android/server/UiServiceTestCase.java
@@ -15,7 +15,9 @@ package com.android.server;
 
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyInt;
+import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.when;
 
@@ -50,11 +52,6 @@ public class UiServiceTestCase {
     protected static final String PKG_P = "com.example.p";
     protected static final String PKG_R = "com.example.r";
 
-    protected static final int UID_N_MR1 = 10001;
-    protected static final int UID_O = 10002;
-    protected static final int UID_P = 10003;
-    protected static final int UID_R = 10004;
-
     @Rule
     public TestableContext mContext =
             spy(new TestableContext(InstrumentationRegistry.getContext(), null));
@@ -63,6 +60,10 @@ public class UiServiceTestCase {
     protected final @UserIdInt int mUserId = UserHandle.getUserId(mUid);
     protected final UserHandle mUser = UserHandle.of(mUserId);
     protected final String mPkg = mContext.getPackageName();
+    protected final int UID_N_MR1 = UserHandle.getUid(mUserId, 1);
+    protected final int UID_O = UserHandle.getUid(mUserId, 2);
+    protected final int UID_P = UserHandle.getUid(mUserId, 3);
+    protected final int UID_R = UserHandle.getUid(mUserId, 4);
 
     protected TestableContext getContext() {
         return mContext;
@@ -93,7 +94,12 @@ public class UiServiceTestCase {
                             return Build.VERSION_CODES.CUR_DEVELOPMENT;
                     }
                 });
-
+        when(mPmi.getPackageUid(eq(PKG_N_MR1), anyLong(), eq(mUserId))).thenReturn(UID_N_MR1);
+        when(mPmi.getPackageUid(eq(PKG_O), anyLong(), eq(mUserId))).thenReturn(UID_O);
+        when(mPmi.getPackageUid(eq(PKG_P), anyLong(), eq(mUserId))).thenReturn(UID_P);
+        when(mPmi.getPackageUid(eq(PKG_R), anyLong(), eq(mUserId))).thenReturn(UID_R);
+        when(mPmi.getPackageUid(eq(mContext.getPackageName()), anyLong(), eq(mUserId)))
+                .thenReturn(mUid);
         LocalServices.removeServiceForTest(UserManagerInternal.class);
         LocalServices.addService(UserManagerInternal.class, mUmi);
         LocalServices.removeServiceForTest(UriGrantsManagerInternal.class);
diff --git a/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java b/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
index d8f72ecc9f4a..3eb35decad73 100644
--- a/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
+++ b/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
@@ -251,7 +251,6 @@ import android.content.pm.IPackageManager;
 import android.content.pm.LauncherApps;
 import android.content.pm.ModuleInfo;
 import android.content.pm.PackageManager;
-import android.content.pm.PackageManagerInternal;
 import android.content.pm.ParceledListSlice;
 import android.content.pm.ResolveInfo;
 import android.content.pm.ShortcutInfo;
@@ -325,6 +324,8 @@ import android.widget.RemoteViews;
 import androidx.test.InstrumentationRegistry;
 import androidx.test.filters.SmallTest;
 
+import com.android.compatibility.common.util.SystemUtil;
+import com.android.compatibility.common.util.ThrowingRunnable;
 import com.android.internal.R;
 import com.android.internal.config.sysui.TestableFlagResolver;
 import com.android.internal.logging.InstanceId;
@@ -377,7 +378,6 @@ import org.mockito.ArgumentMatcher;
 import org.mockito.InOrder;
 import org.mockito.Mock;
 import org.mockito.Mockito;
-import org.mockito.MockitoAnnotations;
 import org.mockito.invocation.InvocationOnMock;
 import org.mockito.stubbing.Answer;
 
@@ -405,7 +405,6 @@ import java.util.function.Consumer;
 @SuppressLint("GuardedBy") // It's ok for this test to access guarded methods from the service.
 public class NotificationManagerServiceTest extends UiServiceTestCase {
     private static final String TEST_CHANNEL_ID = "NotificationManagerServiceTestChannelId";
-    private static final String TEST_PACKAGE = "The.name.is.Package.Test.Package";
     private static final String PKG_NO_CHANNELS = "com.example.no.channels";
     private static final String MISSING_PACKAGE = "MISSING!";
     private static final int TEST_TASK_ID = 1;
@@ -443,8 +442,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @Mock
     private PackageManager mPackageManagerClient;
     @Mock
-    private PackageManagerInternal mPackageManagerInternal;
-    @Mock
     private PermissionPolicyInternal mPermissionPolicyInternal;
     @Mock
     private WindowManagerInternal mWindowManagerInternal;
@@ -610,7 +607,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     }
 
     @Before
-    public void setUp() throws Exception {
+    public void setUpNMS() throws Exception {
         // Shell permisssions will override permissions of our app, so add all necessary permissions
         // for this test here:
         InstrumentationRegistry.getInstrumentation().getUiAutomation().adoptShellPermissionIdentity(
@@ -618,8 +615,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 "android.permission.READ_DEVICE_CONFIG",
                 "android.permission.READ_CONTACTS");
 
-        MockitoAnnotations.initMocks(this);
-
         DeviceIdleInternal deviceIdleInternal = mock(DeviceIdleInternal.class);
         when(deviceIdleInternal.getNotificationAllowlistDuration()).thenReturn(3000L);
 
@@ -637,8 +632,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         LocalServices.addService(ActivityManagerInternal.class, mAmi);
         LocalServices.removeServiceForTest(JobSchedulerInternal.class);
         LocalServices.addService(JobSchedulerInternal.class, mJsi);
-        LocalServices.removeServiceForTest(PackageManagerInternal.class);
-        LocalServices.addService(PackageManagerInternal.class, mPackageManagerInternal);
         LocalServices.removeServiceForTest(PermissionPolicyInternal.class);
         LocalServices.addService(PermissionPolicyInternal.class, mPermissionPolicyInternal);
         LocalServices.removeServiceForTest(ShortcutServiceInternal.class);
@@ -687,14 +680,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         when(mPackageManagerClient.getPackageUidAsUser(any(), anyInt())).thenReturn(mUid);
         when(mPackageManagerClient.getPackageUidAsUser(eq(MISSING_PACKAGE), anyInt()))
                 .thenThrow(new PackageManager.NameNotFoundException("Missing package!"));
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenAnswer(
-                (Answer<Boolean>) invocation -> {
-                    // TODO: b/317957802 - This is overly broad and basically makes ANY
-                    //  isSameApp() check pass,  requiring Mockito.reset() for meaningful
-                    //  tests! Make it more precise.
-                    Object[] args = invocation.getArguments();
-                    return (int) args[1] == mUid;
-                });
         when(mLightsManager.getLight(anyInt())).thenReturn(mock(LogicalLight.class));
         when(mAudioManager.getRingerModeInternal()).thenReturn(AudioManager.RINGER_MODE_NORMAL);
         when(mPackageManagerClient.hasSystemFeature(FEATURE_WATCH)).thenReturn(false);
@@ -945,7 +930,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         assertNotNull(mBinderService.getNotificationChannel(
                 mPkg, mContext.getUserId(), mPkg, mMinChannel.getId()));
         clearInvocations(mRankingHandler);
-        when(mPermissionHelper.hasPermission(mUid)).thenReturn(true);
+        when(mPermissionHelper.hasPermission(anyInt())).thenReturn(true);
 
         var checker = mock(TestableNotificationManagerService.ComponentPermissionChecker.class);
         mService.permissionChecker = checker;
@@ -1154,7 +1139,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     }
 
     private StatusBarNotification generateSbn(String pkg, int uid, long postTime, int userId) {
-        Notification.Builder nb = new Notification.Builder(mContext, "a")
+        Notification.Builder nb = new Notification.Builder(mContext,
+                mTestNotificationChannel.getId())
                 .setContentTitle("foo")
                 .setSmallIcon(android.R.drawable.sym_def_app_icon);
         StatusBarNotification sbn = new StatusBarNotification(pkg, pkg, uid,
@@ -1462,10 +1448,11 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemUid = false;
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
-        setIfPackageHasPermissionToAvoidToastRateLimiting(TEST_PACKAGE, false);
+        setIfPackageHasPermissionToAvoidToastRateLimiting(PKG_P, false);
         // package is not suspended
-        when(mPackageManager.isPackageSuspendedForUser(TEST_PACKAGE, mUserId))
+        when(mPackageManager.isPackageSuspendedForUser(PKG_P, mUserId))
                 .thenReturn(false);
+        when(mPmi.getPackageUid(eq(PKG_P), anyLong(), eq(mUserId))).thenReturn(mUid);
     }
 
     private boolean enqueueToast(String testPackage, ITransientNotification callback)
@@ -1486,7 +1473,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     private boolean enqueueTextToast(String testPackage, CharSequence text, boolean isUiContext,
             int displayId) throws RemoteException {
-        return ((INotificationManager) mService.mService).enqueueTextToast(testPackage,
+        return mBinderService.enqueueTextToast(testPackage,
                 new Binder(), text, TOAST_DURATION, isUiContext, displayId,
                 /* textCallback= */ null);
     }
@@ -1504,7 +1491,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     }
 
     private void verifyToastShownForTestPackage(String text, int displayId) {
-        verify(mStatusBar).showToast(eq(mUid), eq(TEST_PACKAGE), any(), eq(text), any(),
+        verify(mStatusBar).showToast(eq(mUid), eq(PKG_P), any(), eq(text), any(),
                 eq(TOAST_DURATION), any(), eq(displayId));
     }
 
@@ -1668,6 +1655,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @Test
     public void testCreateNotificationChannels_FirstChannelWithFgndTaskStartsPermDialog()
             throws Exception {
+        when(mPmi.getPackageUid(PKG_NO_CHANNELS, 0L, mUserId)).thenReturn(mUid);
         when(mAtm.getTaskToShowPermissionDialogOn(anyString(), anyInt())).thenReturn(TEST_TASK_ID);
         final NotificationChannel channel =
                 new NotificationChannel("id", "name", IMPORTANCE_DEFAULT);
@@ -4572,7 +4560,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         assertThat(mService.mNotificationList).hasSize(2);
         assertThat(mService.mEnqueuedNotifications).hasSize(2);
 
-        when(mPackageManagerInternal.getPackageUid("revoked", 0, 0)).thenReturn(1001);
+        when(mPmi.getPackageUid("revoked", 0L, 0)).thenReturn(1001);
         when(mPermissionHelper.hasPermission(eq(1001))).thenReturn(false);
 
         mOnPermissionChangeListener.onOpChanged(
@@ -4607,7 +4595,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         assertThat(mService.mNotificationList).hasSize(2);
         assertThat(mService.mEnqueuedNotifications).hasSize(2);
 
-        when(mPackageManagerInternal.getPackageUid("granted", 0, 0)).thenReturn(1001);
+        when(mPmi.getPackageUid("granted", 0L, 0)).thenReturn(1001);
         when(mPermissionHelper.hasPermission(eq(1001))).thenReturn(true);
 
         mOnPermissionChangeListener.onOpChanged(
@@ -4626,8 +4614,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 AppOpsManager.OPSTR_POST_NOTIFICATION, "package", 0);
         waitForIdle();
 
-        // We early-exited and didn't even query PM for package details.
-        verify(mPackageManagerInternal, never()).getPackageUid(any(), anyLong(), anyInt());
+        // We early-exited and didn't even query PM for permissions.
+        verify(mPermissionHelper, never()).hasPermission(anyInt());
     }
 
     @Test
@@ -4635,7 +4623,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.addNotification(new NotificationRecord(mContext,
                 generateSbn("package", 1001, 1, 0), mTestNotificationChannel));
         assertThat(mService.mNotificationList).hasSize(1);
-        when(mPackageManagerInternal.getPackageUid("package", 0, 0)).thenReturn(1001);
+        when(mPmi.getPackageUid("package", 0L, 0)).thenReturn(1001);
         when(mPermissionHelper.hasRequestedPermission(any(), eq("package"), anyInt())).thenReturn(
                 true);
 
@@ -7274,6 +7262,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testBumpFGImportance_channelChangePreOApp() throws Exception {
+        when(mPackageManagerClient.getPackageUidAsUser(PKG_N_MR1, mUserId)).thenReturn(UID_N_MR1);
         Notification.Builder nb = new Notification.Builder(mContext,
                 NotificationChannel.DEFAULT_CHANNEL_ID)
                 .setContentTitle("foo")
@@ -7283,11 +7272,11 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
         StatusBarNotification sbn = new StatusBarNotification(PKG_N_MR1, PKG_N_MR1, 9,
                 "testBumpFGImportance_channelChangePreOApp",
-                Binder.getCallingUid(), 0, nb.build(),
-                UserHandle.getUserHandleForUid(Binder.getCallingUid()), null, 0);
+                UID_N_MR1, 0, nb.build(),
+                mUser, null, 0);
 
-        mBinderService.enqueueNotificationWithTag(sbn.getPackageName(), sbn.getOpPkg(),
-                sbn.getTag(), sbn.getId(), sbn.getNotification(), sbn.getUserId());
+        mService.enqueueNotificationInternal(sbn.getPackageName(), sbn.getOpPkg(), UID_N_MR1, 0,
+                sbn.getTag(), sbn.getId(), sbn.getNotification(), sbn.getUserId(), false, true);
         waitForIdle();
 
         assertEquals(IMPORTANCE_LOW,
@@ -7302,19 +7291,24 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 .setPriority(Notification.PRIORITY_MIN);
 
         sbn = new StatusBarNotification(PKG_N_MR1, PKG_N_MR1, 9,
-                "testBumpFGImportance_channelChangePreOApp", Binder.getCallingUid(),
-                0, nb.build(), UserHandle.getUserHandleForUid(Binder.getCallingUid()), null, 0);
+                "testBumpFGImportance_channelChangePreOApp", UID_N_MR1,
+                0, nb.build(), mUser, null, 0);
 
-        mBinderService.enqueueNotificationWithTag(PKG_N_MR1, PKG_N_MR1,
+        mService.enqueueNotificationInternal(PKG_N_MR1, PKG_N_MR1, UID_N_MR1, 0,
                 "testBumpFGImportance_channelChangePreOApp",
-                sbn.getId(), sbn.getNotification(), sbn.getUserId());
+                sbn.getId(), sbn.getNotification(), sbn.getUserId(), false, true);
         waitForIdle();
         assertEquals(IMPORTANCE_LOW,
                 mService.getNotificationRecord(sbn.getKey()).getImportance());
 
-        NotificationChannel defaultChannel = mBinderService.getNotificationChannel(
-                PKG_N_MR1, mContext.getUserId(), PKG_N_MR1, NotificationChannel.DEFAULT_CHANNEL_ID);
-        assertEquals(IMPORTANCE_LOW, defaultChannel.getImportance());
+        SystemUtil.runWithShellPermissionIdentity(
+                () -> {
+                    NotificationChannel defaultChannel = mBinderService.getNotificationChannel(
+                            "shell", mContext.getUserId(), PKG_N_MR1,
+                            NotificationChannel.DEFAULT_CHANNEL_ID);
+                    assertEquals(IMPORTANCE_LOW, defaultChannel.getImportance());
+                }
+        );
     }
 
     @Test
@@ -8896,8 +8890,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.addNotification(notif1);
         mService.addNotification(notif2);
 
-        // on broadcast, nothing is hidden since no notifications are of package "test_package"
-        simulatePackageSuspendBroadcast(true, "test_package", notif1.getUid());
+        // on broadcast, nothing is hidden since no notifications are of package "PKG_P"
+        simulatePackageSuspendBroadcast(true, "PKG_P", notif1.getUid());
         ArgumentCaptor<List> captor = ArgumentCaptor.forClass(List.class);
         verify(mListeners, times(1)).notifyHiddenLocked(captor.capture());
         assertEquals(0, captor.getValue().size());
@@ -8991,9 +8985,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 mTestNotificationChannel, 1, null, true);
         mService.addNotification(notif1);
 
-        // on broadcast, nothing is hidden since no notifications are of package "test_package"
+        // on broadcast, nothing is hidden since no notifications are of package "PKG_P"
         simulatePackageDistractionBroadcast(
-                PackageManager.RESTRICTION_HIDE_NOTIFICATIONS, new String[] {"test_package"},
+                PackageManager.RESTRICTION_HIDE_NOTIFICATIONS, new String[] {"PKG_P"},
                 new int[]{notif1.getUid()});
         ArgumentCaptor<List> captor = ArgumentCaptor.forClass(List.class);
         verify(mListeners, times(1)).notifyHiddenLocked(captor.capture());
@@ -9270,6 +9264,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         info.uid = Binder.getCallingUid();
         when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(10))).thenReturn(info);
         when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(0))).thenReturn(null);
+        when(mPmi.getPackageUid(eq("caller"), anyLong(), eq(mUserId))).thenReturn(info.uid);
+        when(mPmi.getPackageUid(eq("caller"), anyLong(), eq(10))).thenReturn(info.uid);
 
         int actualUid = mService.resolveNotificationUid("caller", "caller", info.uid, 10);
 
@@ -9280,9 +9276,11 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testResolveNotificationUid_sameApp() throws Exception {
         ApplicationInfo info = new ApplicationInfo();
         info.uid = Binder.getCallingUid();
-        when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(0))).thenReturn(info);
+        when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(mUserId)))
+                .thenReturn(info);
+        when(mPmi.getPackageUid(eq("caller"), anyLong(), eq(mUserId))).thenReturn(info.uid);
 
-        int actualUid = mService.resolveNotificationUid("caller", "caller", info.uid, 0);
+        int actualUid = mService.resolveNotificationUid("caller", "caller", info.uid, mUserId);
 
         assertEquals(info.uid, actualUid);
     }
@@ -9291,9 +9289,12 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testResolveNotificationUid_sameAppDiffPackage() throws Exception {
         ApplicationInfo info = new ApplicationInfo();
         info.uid = Binder.getCallingUid();
-        when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(0))).thenReturn(info);
+        when(mPackageManager.getApplicationInfo(anyString(), anyLong(), eq(mUserId)))
+                .thenReturn(info);
+        when(mPmi.getPackageUid(eq("caller"), anyLong(), eq(mUserId))).thenReturn(info.uid);
+        when(mPmi.getPackageUid(eq("callerAlso"), anyLong(), eq(mUserId))).thenReturn(info.uid);
 
-        int actualUid = mService.resolveNotificationUid("caller", "callerAlso", info.uid, 0);
+        int actualUid = mService.resolveNotificationUid("caller", "callerAlso", info.uid, mUserId);
 
         assertEquals(info.uid, actualUid);
     }
@@ -9328,10 +9329,11 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testResolveNotificationUid_androidAllowed() throws Exception {
         int expectedUid = 123;
 
-        when(mPackageManagerClient.getPackageUidAsUser("target", 0)).thenReturn(expectedUid);
+        when(mPackageManagerClient.getPackageUidAsUser("target", mUserId)).thenReturn(expectedUid);
+        when(mPmi.getPackageUid("target", 0, mUserId)).thenReturn(expectedUid);
         // no delegate
 
-        assertEquals(expectedUid, mService.resolveNotificationUid("android", "target", 0, 0));
+        assertEquals(expectedUid, mService.resolveNotificationUid("android", "target", 0, mUserId));
     }
 
     @Test
@@ -9339,6 +9341,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         final String notReal = "NOT REAL";
         when(mPackageManagerClient.getPackageUidAsUser(anyString(), anyInt())).thenThrow(
                 PackageManager.NameNotFoundException.class);
+        when(mPmi.getPackageUid(eq("android"), anyLong(), anyInt())).thenReturn(1000);
         ApplicationInfo ai = new ApplicationInfo();
         ai.uid = -1;
         when(mPackageManager.getApplicationInfo(anyString(), anyLong(), anyInt())).thenReturn(ai);
@@ -9501,6 +9504,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9525,6 +9529,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9546,6 +9551,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9576,6 +9582,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(false); // rate limit reached
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9600,6 +9607,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9635,6 +9643,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9656,6 +9665,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9677,6 +9687,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9709,14 +9720,12 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setToastRateIsWithinQuota(false); // rate limit reached
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
         setAppInForegroundForToasts(mUid, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
                 .thenReturn(false);
 
-        Binder token = new Binder();
-        INotificationManager nmService = (INotificationManager) mService.mService;
-
         enqueueTextToast(testPackage, "Text");
         verify(mStatusBar, times(0))
                 .showToast(anyInt(), any(), any(), any(), any(), anyInt(), any(), anyInt());
@@ -9732,6 +9741,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setToastRateIsWithinQuota(false); // rate limit reached
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
         setAppInForegroundForToasts(mUid, true);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9747,22 +9757,18 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testTextToastRateLimiterAllowsLimitAvoidanceWithPermission() throws Exception {
-        final String testPackage = "testPackageName";
         assertEquals(0, mService.mToastQueue.size());
         mService.isSystemUid = false;
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(false); // rate limit reached
-        setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, true);
+        setIfPackageHasPermissionToAvoidToastRateLimiting(PKG_P, true);
         setAppInForegroundForToasts(mUid, false);
 
         // package is not suspended
-        when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
+        when(mPackageManager.isPackageSuspendedForUser(PKG_P, mUserId))
                 .thenReturn(false);
 
-        Binder token = new Binder();
-        INotificationManager nmService = (INotificationManager) mService.mService;
-
-        enqueueTextToast(testPackage, "Text");
+        enqueueTextToast(mPkg, "Text");
         verify(mStatusBar).showToast(anyInt(), any(), any(), any(), any(), anyInt(), any(),
                 anyInt());
     }
@@ -9776,6 +9782,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setToastRateIsWithinQuota(false); // rate limit reached
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
         setAppInForegroundForToasts(mUid, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9809,6 +9816,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemUid = true;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9835,6 +9843,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9858,6 +9867,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(testPackage, 0L, mUserId)).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -9877,7 +9887,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         allowTestPackageToToast();
 
         // enqueue toast -> no toasts enqueued
-        boolean wasEnqueued = enqueueTextToast(TEST_PACKAGE, "Text");
+        boolean wasEnqueued = enqueueTextToast(PKG_P, "Text");
         assertThat(wasEnqueued).isTrue();
 
         verifyToastShownForTestPackage("Text", DEFAULT_DISPLAY);
@@ -9888,7 +9898,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
             throws Exception {
         allowTestPackageToToast();
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ false, DEFAULT_DISPLAY);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ false, DEFAULT_DISPLAY);
 
         verifyToastShownForTestPackage("Text", DEFAULT_DISPLAY);
     }
@@ -9899,7 +9909,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         allowTestPackageToToast();
         mockIsUserVisible(SECONDARY_DISPLAY_ID, true);
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ false, SECONDARY_DISPLAY_ID);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ false, SECONDARY_DISPLAY_ID);
 
         verifyToastShownForTestPackage("Text", SECONDARY_DISPLAY_ID);
     }
@@ -9911,7 +9921,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mockDisplayAssignedToUser(SECONDARY_DISPLAY_ID);
         allowTestPackageToToast();
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ true, DEFAULT_DISPLAY);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ true, DEFAULT_DISPLAY);
 
         verifyToastShownForTestPackage("Text", DEFAULT_DISPLAY);
 
@@ -9925,7 +9935,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mockDisplayAssignedToUser(INVALID_DISPLAY); // make sure it's not used
         allowTestPackageToToast();
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ true, SECONDARY_DISPLAY_ID);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ true, SECONDARY_DISPLAY_ID);
 
         verifyToastShownForTestPackage("Text", SECONDARY_DISPLAY_ID);
     }
@@ -9938,7 +9948,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mockDisplayAssignedToUser(SECONDARY_DISPLAY_ID);
         allowTestPackageToToast();
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ false, DEFAULT_DISPLAY);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ false, DEFAULT_DISPLAY);
 
         verifyToastShownForTestPackage("Text", SECONDARY_DISPLAY_ID);
     }
@@ -9951,7 +9961,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mockDisplayAssignedToUser(INVALID_DISPLAY); // make sure it's not used
         allowTestPackageToToast();
 
-        enqueueTextToast(TEST_PACKAGE, "Text", /* isUiContext= */ false, SECONDARY_DISPLAY_ID);
+        enqueueTextToast(PKG_P, "Text", /* isUiContext= */ false, SECONDARY_DISPLAY_ID);
 
         verifyToastShownForTestPackage("Text", SECONDARY_DISPLAY_ID);
     }
@@ -9965,6 +9975,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
         mockIsUserVisible(DEFAULT_DISPLAY, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, UserHandle.getUserId(mUid)))
@@ -9986,6 +9997,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -10010,6 +10022,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -10033,6 +10046,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemUid = true;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -10057,6 +10071,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -10086,6 +10101,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackage, false);
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackage, mUserId))
@@ -10104,6 +10120,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackageSystem, false);
         when(mPackageManager.isPackageSuspendedForUser(testPackageSystem, mUserId))
                 .thenReturn(false);
+        when(mPmi.getPackageUid(eq(testPackageSystem), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         enqueueToast(testPackageSystem, new TestableToastCallback());
 
@@ -10119,6 +10136,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemUid = true;
         setToastRateIsWithinQuota(true);
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackageSystem1, false);
+        when(mPmi.getPackageUid(eq(testPackageSystem1), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         // package is not suspended
         when(mPackageManager.isPackageSuspendedForUser(testPackageSystem1, mUserId))
@@ -10137,6 +10155,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         setIfPackageHasPermissionToAvoidToastRateLimiting(testPackageSystem2, false);
         when(mPackageManager.isPackageSuspendedForUser(testPackageSystem2, mUserId))
                 .thenReturn(false);
+        when(mPmi.getPackageUid(eq(testPackageSystem2), anyLong(), eq(mUserId))).thenReturn(mUid);
 
         enqueueToast(testPackageSystem2, new TestableToastCallback());
 
@@ -10522,6 +10541,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testCanNotifyAsUser_crossUser() throws Exception {
+        when(mPmi.getPackageUid("src", 0L, mUserId)).thenReturn(mUid);
+        when(mPmi.getPackageUid("src", 0L, mUserId + 1)).thenReturn(mUid);
         // same user no problem
         mBinderService.canNotifyAsPackage("src", "target", mContext.getUserId());
 
@@ -10540,6 +10561,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testGetNotificationChannels_crossUser() throws Exception {
+        when(mPmi.getPackageUid("src", 0L, mUserId)).thenReturn(mUid);
+        when(mPmi.getPackageUid("src", 0L, mUserId + 1)).thenReturn(mUid);
         // same user no problem
         mBinderService.getNotificationChannels("src", "target", mContext.getUserId());
 
@@ -11142,10 +11165,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @Test
     public void testAddAutomaticZenRule_systemCallTakesPackageFromOwner() throws Exception {
         mService.isSystemUid = true;
-        ZenModeHelper mockZenModeHelper = mock(ZenModeHelper.class);
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
-        mService.setZenHelper(mockZenModeHelper);
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
+        when(mPmi.getPackageUid(eq("com.android.settings"), anyLong(), eq(mUserId)))
+                .thenReturn(mUid);
         ComponentName owner = new ComponentName("android", "ProviderName");
         ZenPolicy zenPolicy = new ZenPolicy.Builder().allowAlarms(true).build();
         boolean isEnabled = true;
@@ -11162,6 +11184,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testAddAutomaticZenRule_systemAppIdCallTakesPackageFromOwner() throws Exception {
         // The multi-user case: where the calling uid doesn't match the system uid, but the calling
         // *appid* is the system.
+        when(mPmi.getPackageUid("com.android.settings", 0L, mUserId)).thenReturn(mUid);
         mService.isSystemUid = false;
         mService.isSystemAppId = true;
         ZenModeHelper mockZenModeHelper = mock(ZenModeHelper.class);
@@ -11184,10 +11207,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testAddAutomaticZenRule_nonSystemCallTakesPackageFromArg() throws Exception {
         mService.isSystemUid = false;
         mService.isSystemAppId = false;
-        ZenModeHelper mockZenModeHelper = mock(ZenModeHelper.class);
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
-        mService.setZenHelper(mockZenModeHelper);
+        when(mPmi.getPackageUid(eq("another.package"), anyLong(), anyInt())).thenReturn(mUid);
+
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
         ComponentName owner = new ComponentName("android", "ProviderName");
         ZenPolicy zenPolicy = new ZenPolicy.Builder().allowAlarms(true).build();
         boolean isEnabled = true;
@@ -11233,8 +11255,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testAddAutomaticZenRule_typeBedtimeCanBeUsedByWellbeing() throws Exception {
         ZenModeHelper zenModeHelper = setUpMockZenTest();
         mService.setCallerIsNormalPackage();
-        reset(mPackageManagerInternal);
-        when(mPackageManagerInternal.isSameApp(eq(mPkg), eq(mUid), anyInt())).thenReturn(true);
         when(mResources
                 .getString(com.android.internal.R.string.config_systemWellbeing))
                 .thenReturn(mPkg);
@@ -11253,15 +11273,11 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testAddAutomaticZenRule_typeBedtimeCanBeUsedBySystem() throws Exception {
-        reset(mPackageManagerInternal);
-        when(mPackageManagerInternal.isSameApp(eq(mPkg), eq(mUid), anyInt())).thenReturn(true);
         addAutomaticZenRule_restrictedRuleTypeCanBeUsedBySystem(AutomaticZenRule.TYPE_BEDTIME);
     }
 
     @Test
     public void testAddAutomaticZenRule_typeBedtimeCannotBeUsedByRegularApps() throws Exception {
-        reset(mPackageManagerInternal);
-        when(mPackageManagerInternal.isSameApp(eq(mPkg), eq(mUid), anyInt())).thenReturn(true);
         addAutomaticZenRule_restrictedRuleTypeCannotBeUsedByRegularApps(
                 AutomaticZenRule.TYPE_BEDTIME);
     }
@@ -11304,6 +11320,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void addAutomaticZenRule_fromUser_mappedToOriginUser() throws Exception {
         ZenModeHelper zenModeHelper = setUpMockZenTest();
         mService.isSystemUid = true;
+        when(mPmi.getPackageUid("pkg", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.addAutomaticZenRule(SOME_ZEN_RULE, "pkg", /* fromUser= */ true);
 
@@ -11315,6 +11332,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void addAutomaticZenRule_fromSystemNotUser_mappedToOriginSystem() throws Exception {
         ZenModeHelper zenModeHelper = setUpMockZenTest();
         mService.isSystemUid = true;
+        when(mPmi.getPackageUid("pkg", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.addAutomaticZenRule(SOME_ZEN_RULE, "pkg", /* fromUser= */ false);
 
@@ -11326,6 +11344,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void addAutomaticZenRule_fromApp_mappedToOriginApp() throws Exception {
         ZenModeHelper zenModeHelper = setUpMockZenTest();
         mService.setCallerIsNormalPackage();
+        when(mPmi.getPackageUid("pkg", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.addAutomaticZenRule(SOME_ZEN_RULE, "pkg", /* fromUser= */ false);
 
@@ -12062,6 +12081,12 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testGrantInlineReplyUriPermission_acrossUsers() throws Exception {
         // generate a NotificationRecord for USER_ALL to make sure it's converted into USER_SYSTEM
         int otherUserId = 11;
+        int uid = 0; // sysui on primary user
+        int otherUserUid = (otherUserId * 100000) + 1; // sysui as a different user
+        String sysuiPackage = "sysui";
+        final String[] sysuiPackages = new String[] { sysuiPackage };
+        when(mPmi.getPackageUid(mContext.getPackageName(), 0L, otherUserId)).thenReturn(mUid);
+
         NotificationRecord nr =
                 generateNotificationRecord(mTestNotificationChannel, otherUserId);
         mBinderService.enqueueNotificationWithTag(mPkg, mPkg, "tag",
@@ -12077,10 +12102,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
         Uri uri = ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, 1);
 
-        int uid = 0; // sysui on primary user
-        int otherUserUid = (otherUserId * 100000) + 1; // sysui as a different user
-        String sysuiPackage = "sysui";
-        final String[] sysuiPackages = new String[] { sysuiPackage };
         when(mPackageManager.getPackagesForUid(uid)).thenReturn(sysuiPackages);
 
         // Make sure to mock call for USER_SYSTEM and not USER_ALL, since it's been replaced by the
@@ -13092,24 +13113,29 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testRecordMessages_invalidMsg() throws RemoteException {
+        mService.mPreferencesHelper.createNotificationChannel(
+                PKG_P, UID_P, mTestNotificationChannel, true, true, UID_P, false);
         Notification.Builder nb = getMessageStyleNotifBuilder(false /* addDefaultMetadata */,
                 null /* groupKey */, false /* isSummary */, true);
         nb.setShortcutId(null);
         StatusBarNotification sbn = new StatusBarNotification(PKG_P, PKG_P, 1,
-                "testRecordMessages_invalidMsg", mUid, 0, nb.build(),
-                UserHandle.getUserHandleForUid(mUid), null, 0);
+                "testRecordMessages_invalidMsg", UID_P, 0, nb.build(),
+                UserHandle.getUserHandleForUid(UID_P), null, 0);
         NotificationRecord nr = new NotificationRecord(mContext, sbn, mTestNotificationChannel);
 
         when(mLauncherApps.getShortcuts(any(), any())).thenReturn(null);
-        mBinderService.enqueueNotificationWithTag(PKG_P, PKG_P, nr.getSbn().getTag(),
-                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId());
+        mService.enqueueNotificationInternal(PKG_P, PKG_P, UID_P, 0, nr.getSbn().getTag(),
+                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId(),
+                false, true);
         waitForIdle();
 
-        assertTrue(mBinderService.isInInvalidMsgState(PKG_P, mUid));
+        assertTrue(mBinderService.isInInvalidMsgState(PKG_P, UID_P));
     }
 
     @Test
     public void testRecordMessages_invalidMsg_notMessageStyle() throws RemoteException {
+        mService.mPreferencesHelper.createNotificationChannel(
+                PKG_O, UID_O, mTestNotificationChannel, true, true, UID_O, false);
         Notification.Builder nb = new Notification.Builder(mContext,
                 mTestNotificationChannel.getId())
                 .setContentTitle("foo")
@@ -13117,44 +13143,50 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 .setSmallIcon(android.R.drawable.sym_def_app_icon)
                 .setCategory(Notification.CATEGORY_MESSAGE);
         StatusBarNotification sbn = new StatusBarNotification(PKG_O, PKG_O, 1,
-                "testRecordMessages_invalidMsg_notMessageStyle", mUid, 0, nb.build(),
-                UserHandle.getUserHandleForUid(mUid), null, 0);
+                "testRecordMessages_invalidMsg_notMessageStyle", UID_O, 0, nb.build(),
+                mUser, null, 0);
         NotificationRecord nr = new NotificationRecord(mContext, sbn, mTestNotificationChannel);
 
         when(mLauncherApps.getShortcuts(any(), any())).thenReturn(null);
-        mBinderService.enqueueNotificationWithTag(PKG_O, PKG_O, nr.getSbn().getTag(),
-                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId());
+        mService.enqueueNotificationInternal(PKG_O, PKG_O, UID_O, 0, nr.getSbn().getTag(),
+                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId(),
+                false, true);
         waitForIdle();
 
         // PKG_O is allowed to be in conversation space b/c of override in
         // TestableNotificationManagerService
-        assertTrue(mBinderService.isInInvalidMsgState(PKG_O, mUid));
+        assertTrue(mBinderService.isInInvalidMsgState(PKG_O, UID_O));
     }
 
     @Test
     public void testRecordMessages_validMsg() throws RemoteException {
+        mService.mPreferencesHelper.createNotificationChannel(
+                PKG_P, UID_P, mTestNotificationChannel, true, true, UID_P, false);
         Notification.Builder nb = getMessageStyleNotifBuilder(false /* addDefaultMetadata */,
                 null /* groupKey */, false /* isSummary */, true);
         nb.setShortcutId(null);
         StatusBarNotification sbn = new StatusBarNotification(PKG_P, PKG_P, 1,
-                "testRecordMessages_validMsg", mUid, 0, nb.build(),
-                UserHandle.getUserHandleForUid(mUid), null, 0);
+                "testRecordMessages_validMsg", UID_P, 0, nb.build(),
+                mUser, null, 0);
         NotificationRecord nr = new NotificationRecord(mContext, sbn, mTestNotificationChannel);
 
-        mBinderService.enqueueNotificationWithTag(PKG_P, PKG_P, nr.getSbn().getTag(),
-                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId());
+        mService.enqueueNotificationInternal(PKG_P, PKG_P, UID_P, 0, nr.getSbn().getTag(),
+                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId(),
+                false, true);
         waitForIdle();
 
-        assertTrue(mBinderService.isInInvalidMsgState(PKG_P, mUid));
+        assertThat(mService.mNotificationsByKey.size()).isEqualTo(1);
+        assertTrue(mBinderService.isInInvalidMsgState(PKG_P, UID_P));
 
         nr = generateMessageBubbleNotifRecord(mTestNotificationChannel,
                 "testRecordMessages_validMsg");
 
-        mBinderService.enqueueNotificationWithTag(PKG_P, PKG_P, nr.getSbn().getTag(),
-                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId());
+        mService.enqueueNotificationInternal(PKG_P, PKG_P, UID_P, 0, nr.getSbn().getTag(),
+                nr.getSbn().getId(), nr.getSbn().getNotification(), nr.getSbn().getUserId(),
+                false, true);
         waitForIdle();
 
-        assertFalse(mBinderService.isInInvalidMsgState(PKG_P, mUid));
+        assertFalse(mBinderService.isInInvalidMsgState(PKG_P, UID_P));
     }
 
     @Test
@@ -14121,28 +14153,49 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     @Test
     public void testGetActiveNotification_filtersUsers() throws Exception {
-        when(mUm.getProfileIds(mUserId, false)).thenReturn(new int[]{mUserId, 10});
+        int userIdProfile = mUserId + 10;
+        int userIdProfileUid = UserHandle.getUid(userIdProfile, mUid);
+        int secondUser = mUserId + 100;
+        int secondUserUid =  UserHandle.getUid(secondUser, mUid);
+
+        when(mUm.getProfileIds(mUserId, false)).thenReturn(new int[]{mUserId, userIdProfile});
+        when(mPmi.getPackageUid(eq(mContext.getPackageName()), anyLong(), eq(userIdProfile)))
+                .thenReturn(userIdProfileUid);
+        when(mPmi.getPackageUid(eq(mContext.getPackageName()), anyLong(), eq(secondUser)))
+                .thenReturn(secondUserUid);
 
         NotificationRecord nr0 =
                 generateNotificationRecord(mTestNotificationChannel, mUserId);
         mBinderService.enqueueNotificationWithTag(mPkg, mPkg, "tag0",
                 nr0.getSbn().getId(), nr0.getSbn().getNotification(), nr0.getSbn().getUserId());
 
-        NotificationRecord nr10 =
-                generateNotificationRecord(mTestNotificationChannel, 10);
-        mBinderService.enqueueNotificationWithTag(mPkg, mPkg, "tag10",
-                nr10.getSbn().getId(), nr10.getSbn().getNotification(), nr10.getSbn().getUserId());
-
-        NotificationRecord nr11 =
-                generateNotificationRecord(mTestNotificationChannel, 11);
-        mBinderService.enqueueNotificationWithTag(mPkg, mPkg, "tag11",
-                nr11.getSbn().getId(), nr11.getSbn().getNotification(), nr11.getSbn().getUserId());
+        SystemUtil.runWithShellPermissionIdentity(
+                () -> {
+                    mBinderService.createNotificationChannelsForPackage(mPkg, userIdProfileUid,
+                            new ParceledListSlice(Arrays.asList(mTestNotificationChannel)));
+                    mBinderService.createNotificationChannelsForPackage(mPkg, secondUserUid,
+                            new ParceledListSlice(Arrays.asList(mTestNotificationChannel)));
+
+                    StatusBarNotification profileSbn = generateSbn(
+                            mPkg, userIdProfileUid, System.currentTimeMillis(), userIdProfile);
+                    mService.enqueueNotificationInternal(profileSbn.getPackageName(),
+                            profileSbn.getOpPkg(),
+                            userIdProfileUid, 0, "tagUserIdProfile", profileSbn.getId(),
+                            profileSbn.getNotification(), profileSbn.getUserId(), false, true);
+
+                    StatusBarNotification user2Sbn = generateSbn(
+                            mPkg, secondUserUid, System.currentTimeMillis(), secondUser);
+                    mService.enqueueNotificationInternal(user2Sbn.getPackageName(),
+                            user2Sbn.getOpPkg(), secondUserUid, 0, "tagSecondUser",
+                            user2Sbn.getId(), user2Sbn.getNotification(), user2Sbn.getUserId(),
+                            false, true);
+                });
         waitForIdle();
 
         StatusBarNotification[] notifs = mBinderService.getActiveNotifications(mPkg);
         assertEquals(2, notifs.length);
         for (StatusBarNotification sbn : notifs) {
-            if (sbn.getUserId() == 11) {
+            if (sbn.getUserId() == secondUser) {
                 fail("leaked data across users");
             }
         }
@@ -14307,9 +14360,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @DisableFlags(FLAG_NOTIFICATION_FORCE_GROUPING)
     public void testUngroupingOngoingAutoSummary() throws Exception {
         NotificationRecord nr0 =
-                generateNotificationRecord(mTestNotificationChannel, 0);
+                generateNotificationRecord(mTestNotificationChannel, mUserId);
         NotificationRecord nr1 =
-                generateNotificationRecord(mTestNotificationChannel, 0);
+                generateNotificationRecord(mTestNotificationChannel, mUserId);
         nr1.getSbn().getNotification().flags |= FLAG_ONGOING_EVENT;
 
         mService.addNotification(nr0);
@@ -14345,9 +14398,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @EnableFlags(FLAG_NOTIFICATION_FORCE_GROUPING)
     public void testUngroupingOngoingAutoSummary_forceGrouping() throws Exception {
         NotificationRecord nr0 =
-            generateNotificationRecord(mTestNotificationChannel, 0);
+            generateNotificationRecord(mTestNotificationChannel, mUserId);
         NotificationRecord nr1 =
-            generateNotificationRecord(mTestNotificationChannel, 0);
+            generateNotificationRecord(mTestNotificationChannel, mUserId);
         nr1.getSbn().getNotification().flags |= FLAG_ONGOING_EVENT;
 
         mService.addNotification(nr0);
@@ -14929,8 +14982,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         timePostedMs -= BITMAP_DURATION.toMillis();
 
         when(mPermissionHelper.hasPermission(UID_O)).thenReturn(true);
-        when(mPackageManagerInternal.isSameApp(PKG_O, UID_O, UserHandle.getUserId(UID_O)))
-                .thenReturn(true);
         mService.mPreferencesHelper.createNotificationChannel(PKG_O, UID_O,
                 mTestNotificationChannel, true /* fromTargetApp */, false, UID_O,
                 false);
@@ -16825,10 +16876,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @EnableCompatChanges(NotificationManagerService.MANAGE_GLOBAL_ZEN_VIA_IMPLICIT_RULES)
     public void setNotificationPolicy_mappedToImplicitRule() throws RemoteException {
         mService.setCallerIsNormalPackage();
-        ZenModeHelper zenHelper = mock(ZenModeHelper.class);
-        mService.mZenModeHelper = zenHelper;
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
+
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
 
         NotificationManager.Policy policy = new NotificationManager.Policy(0, 0, 0);
         mBinderService.setNotificationPolicy("package", policy, false);
@@ -16845,6 +16895,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
                 .thenReturn(true);
         mService.isSystemUid = true;
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         NotificationManager.Policy policy = new NotificationManager.Policy(0, 0, 0);
         mBinderService.setNotificationPolicy("package", policy, false);
@@ -16890,6 +16941,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                                 .setDisplayName("My connected device")
                                 .setDeviceProfile(deviceProfile)
                                 .build()));
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         NotificationManager.Policy policy = new NotificationManager.Policy(0, 0, 0);
         mBinderService.setNotificationPolicy("package", policy, false);
@@ -16906,11 +16958,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @DisableCompatChanges(NotificationManagerService.MANAGE_GLOBAL_ZEN_VIA_IMPLICIT_RULES)
     public void setNotificationPolicy_withoutCompat_setsGlobalPolicy() throws RemoteException {
         mService.setCallerIsNormalPackage();
-        ZenModeHelper zenModeHelper = mock(ZenModeHelper.class);
-        mService.mZenModeHelper = zenModeHelper;
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
-
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
         NotificationManager.Policy policy = new NotificationManager.Policy(0, 0, 0);
         mBinderService.setNotificationPolicy("package", policy, false);
 
@@ -16921,10 +16970,9 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @EnableCompatChanges(NotificationManagerService.MANAGE_GLOBAL_ZEN_VIA_IMPLICIT_RULES)
     public void getNotificationPolicy_mappedFromImplicitRule() throws RemoteException {
         mService.setCallerIsNormalPackage();
-        ZenModeHelper zenHelper = mock(ZenModeHelper.class);
-        mService.mZenModeHelper = zenHelper;
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
+
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.getNotificationPolicy("package");
 
@@ -16935,10 +16983,10 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @EnableCompatChanges(NotificationManagerService.MANAGE_GLOBAL_ZEN_VIA_IMPLICIT_RULES)
     public void setInterruptionFilter_mappedToImplicitRule() throws RemoteException {
         mService.setCallerIsNormalPackage();
-        ZenModeHelper zenHelper = mock(ZenModeHelper.class);
-        mService.mZenModeHelper = zenHelper;
-        when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
-                .thenReturn(true);
+
+        ZenModeHelper zenModeHelper = setUpMockZenTest();
+
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.setInterruptionFilter("package", INTERRUPTION_FILTER_PRIORITY, false);
 
@@ -16955,6 +17003,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         when(mConditionProviders.isPackageOrComponentAllowed(anyString(), anyInt()))
                 .thenReturn(true);
         mService.isSystemUid = true;
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.setInterruptionFilter("package", INTERRUPTION_FILTER_PRIORITY, false);
 
@@ -16998,6 +17047,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                                 .setDisplayName("My connected device")
                                 .setDeviceProfile(deviceProfile)
                                 .build()));
+        when(mPmi.getPackageUid("package", 0L, mUserId)).thenReturn(mUid);
 
         mBinderService.setInterruptionFilter("package", INTERRUPTION_FILTER_PRIORITY, false);
 
@@ -17432,6 +17482,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         waitForIdle();
 
         final UserHandle otherUser = UserHandle.of(2);
+        when(mPmi.getPackageUid(eq(mContext.getPackageName()), anyLong(), eq(2)))
+                .thenReturn(mUid);
         final NotificationRecord r = createAndPostCallStyleNotification(mPkg,
                 otherUser, "testCallNotificationListener_registerForUserAll_notifiedOnAnyUserId");
 
@@ -17676,7 +17728,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
         when(mPackageManagerClient.getPackageUidAsUser(anyString(), anyInt()))
                 .thenThrow(PackageManager.NameNotFoundException.class);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(false);
+        when(mPmi.getPackageUid(anyString(), anyLong(), anyInt())).thenReturn(INVALID_UID);
 
         mInternalService.cancelNotification(mPkg, mPkg, mUid, 0, r.getSbn().getTag(),
                 r.getSbn().getId(), mUserId);
@@ -17691,7 +17743,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
         when(mPackageManagerClient.getPackageUidAsUser(anyString(), anyInt()))
                 .thenThrow(PackageManager.NameNotFoundException.class);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(false);
+        when(mPmi.getPackageUid(anyString(), anyLong(), anyInt())).thenReturn(INVALID_UID);
 
         try {
             mBinderService.enqueueNotificationWithTag(mPkg, mPkg, r.getSbn().getTag(),
@@ -17707,7 +17759,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     public void testGetEffectsSuppressor_noSuppressor() throws Exception {
         when(mUmInternal.getProfileIds(anyInt(), anyBoolean())).thenReturn(new int[]{mUserId});
         when(mListeners.checkServiceTokenLocked(any())).thenReturn(mListener);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(true);
         assertThat(mBinderService.getEffectsSuppressor()).isNull();
     }
 
@@ -17719,7 +17770,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         mBinderService.requestHintsFromListener(mock(INotificationListener.class),
                 HINT_HOST_DISABLE_EFFECTS);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(true);
         assertThat(mBinderService.getEffectsSuppressor()).isEqualTo(mListener.component);
     }
 
@@ -17731,7 +17781,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemAppId = false;
         mBinderService.requestHintsFromListener(mock(INotificationListener.class),
                 HINT_HOST_DISABLE_EFFECTS);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(false);
+        when(mPmi.getPackageUid(anyString(), anyLong(), anyInt())).thenReturn(INVALID_UID);
         assertThat(mBinderService.getEffectsSuppressor()).isEqualTo(null);
     }
 
@@ -17742,7 +17792,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         mService.isSystemUid = true;
         mBinderService.requestHintsFromListener(mock(INotificationListener.class),
                 HINT_HOST_DISABLE_EFFECTS);
-        when(mPackageManagerInternal.isSameApp(anyString(), anyInt(), anyInt())).thenReturn(false);
+        when(mPmi.getPackageUid(anyString(), anyLong(), anyInt())).thenReturn(INVALID_UID);
         assertThat(mBinderService.getEffectsSuppressor()).isEqualTo(mListener.component);
     }
 
@@ -18983,6 +19033,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @Test
     public void onDisplayRemoveSystemDecorations_cancelToasts() throws RemoteException {
         final String testPackage = "testPackageName";
+        when(mPmi.getPackageUid(eq(testPackage), anyLong(), eq(mUserId))).thenReturn(mUid);
         final INotificationManager service = ((INotificationManager) mService.mService);
         final IBinder firstExternal = new Binder();
         final IBinder secondExternal = new Binder();
diff --git a/services/tests/uiservicestests/src/com/android/server/notification/PreferencesHelperTest.java b/services/tests/uiservicestests/src/com/android/server/notification/PreferencesHelperTest.java
index e159fc8d05ef..49b1c59c67a6 100644
--- a/services/tests/uiservicestests/src/com/android/server/notification/PreferencesHelperTest.java
+++ b/services/tests/uiservicestests/src/com/android/server/notification/PreferencesHelperTest.java
@@ -1227,19 +1227,21 @@ public class PreferencesHelperTest extends UiServiceTestCase {
         ByteArrayOutputStream baos = writeXmlAndPurge(
                 PKG_N_MR1, UID_N_MR1, false, USER_SYSTEM);
         String expected = "<ranking version=\"4\" "
-                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n"
-                + "<package name=\"com.example.o\" show_badge=\"true\" "
+                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n";
+        String expected_o =
+                "<package name=\"com.example.o\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
-                + "=\"false\" uid=\"10002\">\n"
+                + "=\"false\" uid=\"" + UID_O + "\">\n"
                 + "<channel id=\"id\" name=\"name\" importance=\"2\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" orig_imp=\"2\" />\n"
-                + "</package>\n"
-                + "<package name=\"com.example.n_mr1\" show_badge=\"true\" "
+                + "</package>\n";
+        String expected_n =
+                "<package name=\"com.example.n_mr1\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
-                + "=\"false\" uid=\"10001\">\n"
+                + "=\"false\" uid=\"" + UID_N_MR1 + "\">\n"
                 + "<channelGroup id=\"1\" name=\"bye\" blocked=\"false\" locked=\"0\" />\n"
                 + "<channelGroup id=\"2\" name=\"hello\" blocked=\"false\" locked=\"0\" />\n"
                 + "<channel id=\"id1\" name=\"name1\" importance=\"4\" show_badge=\"true\" "
@@ -1254,11 +1256,16 @@ public class PreferencesHelperTest extends UiServiceTestCase {
                 + "<channel id=\"miscellaneous\" name=\"Uncategorized\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" />\n"
-                + "</package>\n"
-                + "<package name=\"com.example.p\" show_badge=\"true\" "
+                + "</package>\n";
+        String expected_p =
+                "<package name=\"com.example.p\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"true\" sent_valid_msg=\"true\""
-                + " user_demote_msg_app=\"true\" sent_valid_bubble=\"false\" uid=\"10003\"";
-        assertThat(baos.toString()).contains(expected);
+                + " user_demote_msg_app=\"true\" sent_valid_bubble=\"false\" uid=\""+ UID_P + "\" />";
+        String actual = baos.toString();
+        assertThat(actual).contains(expected);
+        assertThat(actual).contains(expected_o);
+        assertThat(actual).contains(expected_n);
+        assertThat(actual).contains(expected_p);
     }
 
     @Test
@@ -1312,18 +1319,20 @@ public class PreferencesHelperTest extends UiServiceTestCase {
         ByteArrayOutputStream baos = writeXmlAndPurge(
                 PKG_N_MR1, UID_N_MR1, true, USER_SYSTEM);
         String expected = "<ranking version=\"4\" "
-                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n"
+                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n";
+        String expected_o =
                 // Importance 0 because off in permissionhelper
-                + "<package name=\"com.example.o\" importance=\"0\" show_badge=\"true\" "
+                "<package name=\"com.example.o\" importance=\"0\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
                 + "=\"false\">\n"
                 + "<channel id=\"id\" name=\"name\" importance=\"2\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" orig_imp=\"2\" />\n"
-                + "</package>\n"
+                + "</package>\n";
+        String expected_n =
                 // Importance default because on in permission helper
-                + "<package name=\"com.example.n_mr1\" importance=\"3\" show_badge=\"true\" "
+                "<package name=\"com.example.n_mr1\" importance=\"3\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
                 + "=\"false\">\n"
@@ -1341,15 +1350,20 @@ public class PreferencesHelperTest extends UiServiceTestCase {
                 + "<channel id=\"miscellaneous\" name=\"Uncategorized\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" />\n"
-                + "</package>\n"
+                + "</package>\n";
+        String expected_p =
                 // Importance default because on in permission helper
-                + "<package name=\"com.example.p\" importance=\"3\" show_badge=\"true\" "
+                "<package name=\"com.example.p\" importance=\"3\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"true\" sent_valid_msg=\"true\""
                 + " user_demote_msg_app=\"true\" sent_valid_bubble=\"false\"";
-        assertThat(baos.toString()).contains(expected);
+        String actual = baos.toString();
+        assertThat(actual).contains(expected);
+        assertThat(actual).contains(expected_o);
+        assertThat(actual).contains(expected_n);
+        assertThat(actual).contains(expected_p);
         // Packages that exist solely in permissionhelper
-        assertThat(baos.toString()).contains("<package name=\"first\" importance=\"3\"");
-        assertThat(baos.toString()).contains("<package name=\"third\" importance=\"0\"");
+        assertThat(actual).contains("<package name=\"first\" importance=\"3\"");
+        assertThat(actual).contains("<package name=\"third\" importance=\"0\"");
     }
 
     @Test
@@ -1389,7 +1403,7 @@ public class PreferencesHelperTest extends UiServiceTestCase {
         mHelper.createNotificationChannel(PKG_N_MR1, UID_N_MR1, channel3, false, false,
                 SYSTEM_UID, true);
         mHelper.createNotificationChannel(PKG_O, UID_O, getChannel(), true, false,
-                UID_N_MR1, false);
+                UID_O, false);
 
         mHelper.setShowBadge(PKG_N_MR1, UID_N_MR1, true);
         mHelper.setInvalidMessageSent(PKG_P, UID_P);
@@ -1399,18 +1413,20 @@ public class PreferencesHelperTest extends UiServiceTestCase {
         ByteArrayOutputStream baos = writeXmlAndPurge(
                 PKG_N_MR1, UID_N_MR1, true, USER_SYSTEM);
         String expected = "<ranking version=\"4\" "
-                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n"
+                + "last_bubbles_version_upgrade=\"" + Build.VERSION.SDK_INT + "\">\n";
                 // Importance 0 because off in permissionhelper
-                + "<package name=\"com.example.o\" importance=\"0\" show_badge=\"true\" "
+        String expected_o =
+                "<package name=\"com.example.o\" importance=\"0\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
                 + "=\"false\">\n"
                 + "<channel id=\"id\" name=\"name\" importance=\"2\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" orig_imp=\"2\" />\n"
-                + "</package>\n"
-                // Importance 0 because missing from permission helper
-                + "<package name=\"com.example.n_mr1\" importance=\"0\" show_badge=\"true\" "
+                + "</package>\n";
+        String expected_n =
+                // Importance missing because missing from permission helper
+                "<package name=\"com.example.n_mr1\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"false\" "
                 + "sent_valid_msg=\"false\" user_demote_msg_app=\"false\" sent_valid_bubble"
                 + "=\"false\">\n"
@@ -1428,12 +1444,17 @@ public class PreferencesHelperTest extends UiServiceTestCase {
                 + "<channel id=\"miscellaneous\" name=\"Uncategorized\" "
                 + "sound=\"content://settings/system/notification_sound\" usage=\"5\" "
                 + "content_type=\"4\" flags=\"0\" show_badge=\"true\" />\n"
-                + "</package>\n"
+                + "</package>\n";
+        String expected_p =
                 // Importance default because on in permission helper
-                + "<package name=\"com.example.p\" importance=\"3\" show_badge=\"true\" "
+                "<package name=\"com.example.p\" importance=\"3\" show_badge=\"true\" "
                 + "app_user_locked_fields=\"0\" sent_invalid_msg=\"true\" sent_valid_msg=\"true\""
                 + " user_demote_msg_app=\"true\" sent_valid_bubble=\"false\"";
-        assertThat(baos.toString()).contains(expected);
+        String actual = baos.toString();
+        assertThat(actual).contains(expected);
+        assertThat(actual).contains(expected_o);
+        assertThat(actual).contains(expected_n);
+        assertThat(actual).contains(expected_p);
     }
 
     @Test
@@ -4016,7 +4037,7 @@ public class PreferencesHelperTest extends UiServiceTestCase {
         String dump = dumpToString(mHelper);
 
         assertThat(dump).contains(
-                "AppSettings: com.example.p (10003)\n"
+                "AppSettings: com.example.p (" + UID_P + ")\n"
                 + "    Delegate: the.delegate.package (456) enabled=true");
     }
 
-- 
2.53.0

